### ISSUE TITLE
Using APEX on multinode jobs when HPX_WITH_NETWORKING = OFF

### ISSUE THREAD

**Issue by biddisco** [STEllAR-GROUP/hpx#4308]
As documented in #4257 
https://github.com/STEllAR-GROUP/hpx/pull/4257#issuecomment-575370006
and 
https://github.com/STEllAR-GROUP/hpx/pull/4257#issuecomment-575547782
I would like to use apex with user MPI driven jobs. I have modified the init of APEX to use MPI with an experimental (not cleaned up) version looking like 
```
#ifdef HPX_HAVE_APEX
#if !defined(HPX_HAVE_NETWORKING) && defined(HPX_HAVE_LIB_MPI)
        int rank_=0, size_=1, is_initialized_=0;
        // Check if MPI_Init has been called previously
        MPI_Initialized(&is_initialized_);
        if(is_initialized_) {
            MPI_Comm_rank(MPI_COMM_WORLD, &rank_);
            MPI_Comm_size(MPI_COMM_WORLD, &size_);
        }
        else {
            int required, provided;
            required = MPI_THREAD_MULTIPLE;
            MPI_Init_thread(0, nullptr, required, &provided);
            if (provided < MPI_THREAD_FUNNELED) {
                std::cout << "Your MPI installation doesn't allow multiple threads. Exiting.\n";
                std::terminate();
            }
            MPI_Comm_rank(MPI_COMM_WORLD, &rank_);
            MPI_Comm_size(MPI_COMM_WORLD, &size_);
            std::cout << "Initializing mpi on rank " << rank_ << " of " << size_ << std::endl;
        }
        std::cout << "Initializing instrumentation with rank " << rank_ << " of " << size_ << std::endl;
        util::external_timer::init(nullptr, rank_, size_);
#else
        util::external_timer::init(nullptr, hpx::get_locality_id(),
                hpx::get_initial_num_localities());
#endif
```
I have enabled debugging in OTF2 and I see output from my test
```
Initializing mpi on rank 0 of 2                                                                                                                                                               
Initializing instrumentation with rank 0 of 2                                                                                                                                                 
Initializing mpi on rank 1 of 2                                                                                                                                                               
Initializing instrumentation with rank 1 of 2                                                                                                                                                 
Rank 0 of 2.                                                                                                                                                                                  
Rank 1 of 2.                                                                                                                                                                                  
[OTF2] src/OTF2_Archive.c:93: Entering function 'OTF2_Archive_Open': OTF2_archive/APEX                                                                                                        
[OTF2] Active debug module(s): ARCHIVE READER ANCHOR_FILE COLLECTIVES POSIX SION SION_RANK_MAP SION_COLLECTIVES LOCKS                                                                         
[OTF2] Active debug module(s): ARCHIVE READER ANCHOR_FILE COLLECTIVES POSIX SION SION_RANK_MAP SION_COLLECTIVES LOCKS                                                                         
[OTF2] src/OTF2_Archive.c:93: Entering function 'OTF2_Archive_Open': OTF2_archive/APEX                                                                                                        
[OTF2] src/otf2_collectives.c:102: Entering function 'otf2_collectives_get_rank'                                                                                                              
[OTF2] src/otf2_collectives.c:102: Entering function 'otf2_collectives_get_rank'                                                                                                              
[OTF2] src/otf2_collectives.c:181: Entering function 'otf2_collectives_bcast': (nil), 0x7ffecb0440c0, 1, INT64, 0                                                                             
Rank 1 Fudged my_OTF2Bcast                                                                                                                                                                    
[OTF2] src/otf2_collectives.c:181: Entering function 'otf2_collectives_bcast': (nil), 0x7ffcbbf858f0, 1, INT64, 0                                                                             
Rank 0 Fudged my_OTF2Bcast                                                                                                                                                                    

Elapsed time: 0.0599711 seconds                                                                                                                                                               
Cores detected: 8                                                                                                                                                                             
Worker Threads observed: 4                                                                                                                                                                    
Available CPU time: 0.239884 seconds                                                                                                                                                          

Counter                                   : #samples | minimum |    mean  |  maximum |  stddev                                                                                                
------------------------------------------------------------------------------------------------                                                                                              
                    1 Minute Load average :        1      1.100      1.100      1.100      0.000                                                                                              
------------------------------------------------------------------------------------------------                                                                                              

Timer                                                : #calls  |    mean  |   total  |  % total  
------------------------------------------------------------------------------------------------
                                           <unknown> :        1      0.006      0.006      2.666
                                       Dataflow Task :        1      0.001      0.001      0.444
hpx::components::server::runtime_support::shutdow... :        1      0.006      0.006      2.650
hpx::components::server::runtime_support::load_co... :        1      0.006      0.006      2.484
hpx::components::server::runtime_support::call_sh... :        2      0.001      0.003      1.176
hpx::agas::server::primary_namespace::colocate_ac... :        2      0.000      0.000      0.083
hpx::actions::action<void (*)(hpx::components::se... :        2      0.000      0.000      0.129
                        async_launch_policy_dispatch :        3      0.000      0.000      0.073
                                           backtrace :        1      0.001      0.001      0.351
   hpx::parallel::execution::parallel_executor::post :        3      0.000      0.001      0.476
                                          run_helper :        1      0.010      0.010      3.988
                                           APEX Idle :                          0.205     85.480
------------------------------------------------------------------------------------------------
                                        Total timers : 18
Closing OTF2 event files...
Writing OTF2 definition files...
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967296.def'
Writing OTF2 Global definition file...
[OTF2] src/otf2_collectives.c:102: Entering function 'otf2_collectives_get_rank'
Writing OTF2 Node information...
Writing OTF2 Communicators...
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/0.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967297.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/1.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967298.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/2.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967299.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/3.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967300.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4.def'
[OTF2] src/otf2_collectives.c:102: Entering function 'otf2_collectives_get_rank'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967300.evt'
Closing the archive...
[OTF2] src/otf2_collectives.c:102: Entering function 'otf2_collectives_get_rank'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX.otf2'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967299.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/3.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967298.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/2.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967297.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/1.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4294967296.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/0.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX.def'
done.
```
but when loaded in vampir, I do not see "Dataflow Task" for example - there may be other tasks missing, but I didn't look for them. The trace shows two ranks, each with 4 threads, but not all tasks are present, or labelled incorrectly?
When I run the same job with mpi ranks =1, then I see the correct task names.

What other functions must I implement to get this to work?

This is a screenshot of the output from a mpi rank node run
```
rm -rf OTF2_archive && ~/apps/mpich/3.3.1/bin/mpiexec -n 1 bin/scheduler_priority_check_test                                                                        
Initializing mpi on rank 0 of 1                                                                                                                                                               
Initializing instrumentation with rank 0 of 1                                                                                                                                                 
Rank 0 of 1.                                                                                                                                                                                  
[OTF2] src/OTF2_Archive.c:93: Entering function 'OTF2_Archive_Open': OTF2_archive/APEX                                                                                                        
[OTF2] Active debug module(s): ARCHIVE READER ANCHOR_FILE COLLECTIVES POSIX SION SION_RANK_MAP SION_COLLECTIVES LOCKS                                                                         
[OTF2] src/otf2_collectives.c:102: Entering function 'otf2_collectives_get_rank'                                                                                                              
[OTF2] src/otf2_collectives.c:181: Entering function 'otf2_collectives_bcast': (nil), 0x7ffc9365bb50, 1, INT64, 0                                                                             
Rank 0 Fudged my_OTF2Bcast                                                                                                                                                                    

Elapsed time: 0.029733 seconds
Cores detected: 8
Worker Threads observed: 4
Available CPU time: 0.118932 seconds

Timer                                                : #calls  |    mean  |   total  |  % total  
------------------------------------------------------------------------------------------------
                                           <unknown> :        1      0.001      0.001      0.922
                                       Dataflow Task :        1      0.001      0.001      0.882
hpx::components::server::runtime_support::shutdow... :        1      0.001      0.001      0.840
hpx::components::server::runtime_support::load_co... :        1      0.006      0.006      5.153
hpx::components::server::runtime_support::call_sh... :        2      0.001      0.003      2.433
hpx::agas::server::primary_namespace::colocate_ac... :        2      0.000      0.000      0.152
hpx::actions::action<void (*)(hpx::components::se... :        2      0.000      0.000      0.248
                        async_launch_policy_dispatch :        3      0.000      0.000      0.144
                                           backtrace :        1      0.001      0.001      0.556
   hpx::parallel::execution::parallel_executor::post :        3      0.000      0.001      0.951
                                          run_helper :        1      0.010      0.010      8.519
                                           APEX Idle :                          0.094     79.198
------------------------------------------------------------------------------------------------
                                        Total timers : 18
Closing OTF2 event files...
Writing OTF2 definition files...
Writing OTF2 Global definition file...
[OTF2] src/otf2_collectives.c:102: Entering function 'otf2_collectives_get_rank'
Writing OTF2 Node information...
Writing OTF2 Communicators...
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/0.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/1.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/2.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/3.def'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4.def'
Closing the archive...
[OTF2] src/otf2_collectives.c:102: Entering function 'otf2_collectives_get_rank'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX.otf2'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/4.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/3.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/2.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/1.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX/0.evt'
[OTF2] src/otf2_file_posix.c:141: File to open: 'OTF2_archive/APEX.def'
done.
```
![image](https://user-images.githubusercontent.com/941548/72601964-b8c1e500-3916-11ea-9cc4-240ae1ac676b.png)

and this is the vampir view of the two rank run, there is not DataFlow task (among others missing)
![image](https://user-images.githubusercontent.com/941548/72601846-7bf5ee00-3916-11ea-9401-050dec2e1805.png)

**Comment by biddisco** [comment]
Just an additional comment
If I run my test with `--hpx:run-hpx-main` and N>1 ranks, I'd expect to see the task counts for each test task increase since they are being run on all ranks. However in this example of 4 nodes, I se only 1 of each in the summary, so apex is not summing the numbers correctly in the output

```
rm -rf OTF2_archive && ~/apps/mpich/3.3.1/bin/mpiexec -n 4 bin/scheduler_priority_check_test --hpx:run-hpx-main                                                     
Initializing mpi on rank 0 of 4                                                                                                                                                               
Initializing instrumentation with rank 0 of 4                                                                                                                                                 
Initializing mpi on rank 1 of 4                                                                                                                                                               
Initializing instrumentation with rank 1 of 4                                                                                                                                                 
Initializing mpi on rank 3 of 4                                                                                                                                                               
Initializing instrumentation with rank 3 of 4                                                                                                                                                 
...                                                                                                                                                             
 
Elapsed time: 0.174073 seconds                                                                                                                                                                
Cores detected: 8                                                                                                                                                                             
Worker Threads observed: 4                                                                                                                                                                    
Available CPU time: 0.696293 seconds                                                                                                                                                          

Timer                                                : #calls  |    mean  |   total  |  % total                                                                                               
------------------------------------------------------------------------------------------------                                                                                              
                          Continuation with executor :        1      0.004      0.004      0.572                                                                                              
                       Continuation without executor :        1      0.004      0.004      0.576                                                                                              
                              Dataflow with executor :        1      0.003      0.003      0.386                                                                                              
                           Dataflow without executor :        1      0.008      0.008      1.135                                                                                              
hpx::components::server::runtime_support::shutdow... :        1      0.005      0.005      0.681                                                                                              
hpx::components::server::runtime_support::load_co... :        1      0.024      0.024      3.501                                                                                              
hpx::components::server::runtime_support::call_sh... :        4      0.001      0.003      0.448                                                                                              
hpx::agas::server::primary_namespace::colocate_ac... :        2      0.000      0.000      0.029                                                                                              
hpx::actions::action<void (*)(hpx::components::se... :        2      0.000      0.000      0.038                                                                                              
                        async_launch_policy_dispatch :        3      0.000      0.000      0.026                                                                                              
                                          run_helper :        1      0.020      0.020      2.878                                                                                              
                                           APEX Idle :                          0.625     89.730                                                                                              
------------------------------------------------------------------------------------------------         
```
NB. I'm running N ranks on the same node (laptop) here. I hope that isn't the cause of my trouble.

**Comment by khuck** [comment]
The screen output is misleading, it only shows data from rank 0.  I never got around to figuring out how to use HPX to reduce the data at the end of execution.  That’s on the todo list.  It’s the oldest open issue in the list: https://github.com/khuck/xpress-apex/issues/10

In order for OTF2 output to work correctly, the ranks need to know what their rank is.  That happens at `apex::init()` by passing in the rank and size.  If HPX doesn’t know, then APEX doesn’t know, and all ranks assume they are 0.

**Comment by biddisco** [comment]
At the top of this issue, you will see that I have initialized the apex layer with the correct MPI rank information. I can verify that OTF2 knows it is part of an MPI job. It gets the correct ranks. 
Has apex ever been test with HPX on multi rank jobs using normal (network enabled) HPX? I'm no longer convinced that it will work...

**Comment by khuck** [comment]
> At the top of this issue, you will see that I have initialized the apex layer with the correct MPI rank information. I can verify that OTF2 knows it is part of an MPI job. It gets the correct ranks.
> Has apex ever been test with HPX on multi rank jobs using normal (network enabled) HPX? I'm no longer convinced that it will work...

Yes, multi-rank HPX with APEX has been tested many times.  This particular use case hasn't been tested before.  If HPX has networking disabled, and APEX is depending on HPX for networking information, how could it work?...

That said, there's no reason why tasks should be missing, unless they are annotated on one node and executed on another, and the annotation isn't getting transmitted?  Another possibility is that without MPI present, APEX is depending on the filesystem to do event unification (which has been thoroughly tested, and was a pain.  I don't like doing it that way, and it requires a shared filesystem, but it was a temporary fix until I could add the HPX implementation...which hasn't happened).  If you compile APEX with APEX_HAVE_MPI=1 defined, it will use MPI to reduce the global metadata in OTF (in fact, that's the only use of that macro definition).  See https://github.com/khuck/xpress-apex/blob/e4ef45713ac2b225ced19727abd58ab5327cead8/src/apex/otf2_listener.cpp#L1146.  We could also add support in the otf2_listener.cpp (or apex.cpp during initialization) to check for this exact case, where APEX is configured with HPX support but networking is disabled, in which case APEX could be configured to rely on MPI to get the rank/comm_size numbers as well as handle the event unification.

**Comment by biddisco** [comment]
> Yes, multi-rank HPX with APEX has been tested many times. This particular use case hasn't been tested before. If HPX has networking disabled, and APEX is depending on HPX for networking information, how could it work?...

Well that's what this issue is asking. In this comment, https://github.com/STEllAR-GROUP/hpx/pull/4257#issuecomment-563333832 I asked you what I needed to implement to get apex working with MPI only and no HPX networking and you told me that apex doesn't use any HPX networking features. It only requires the right MPI info at init time. I've implemented that, but it doesn't work, so my question is - what else do I need to implement?

You  say 
 
>  Another possibility is that without MPI present, 

but MPI is present. (OTF2 has many callbacks for communication which are not implemented in HPX  - in the otf2 listener - matter?).  

>If you compile APEX with APEX_HAVE_MPI=1 defined, it will use MPI to reduce the global metadata in OTF (in fact, that's the only use of that macro definition). See https://github.com/khuck/xpress-apex/blob/e4ef45713ac2b225ced19727abd58ab5327cead8/src/apex/otf2_listener.cpp#L1146. We could also add support in the otf2_listener.cpp (or apex.cpp during initialization) to check for this exact case, where APEX is configured with HPX support but networking is disabled, in which case APEX could be configured to rely on MPI to get the rank/comm_size numbers as well as handle the event unification.

OK. that starts to make sense. I will look into the `APEX_HAVE_MPI=1` case. This must be the missing part of my setup. Thanks

**Comment by biddisco** [comment]
I've verified that apex is performing a reduction of the otf2 regions at the end of the program, I added some printfs to see the regions in the reduced region map and I can indeed see everything that I expect to see. doing a grep on the apex generated files reveals that the functions that are missing from vampir are present in the files (though that only tells me that otf2 wrote something out with the names in). However vampir does not show the tasks.
The output on my console is
```
Timer                                                : #calls  |    mean  |   total  |  % total  
------------------------------------------------------------------------------------------------
                          Continuation with executor :        1      0.008      0.008      3.213
                       Continuation without executor :        1      0.004      0.004      1.579
                              Dataflow with executor :        1      0.001      0.001      0.445
                           Dataflow without executor :        1      0.004      0.004      1.436
hpx::components::server::runtime_support::shutdow... :        1      0.001      0.001      0.260
hpx::components::server::runtime_support::load_co... :        1      0.006      0.006      2.497
hpx::components::server::runtime_support::call_sh... :        4      0.004      0.015      6.023
hpx::agas::server::primary_namespace::colocate_ac... :        2      0.000      0.000      0.075
hpx::actions::action<void (*)(hpx::components::se... :        2      0.000      0.000      0.107
                        async_launch_policy_dispatch :        3      0.000      0.000      0.065
                                          run_helper :        1      0.024      0.024      9.627
                                           APEX Idle :                          0.185     74.673
------------------------------------------------------------------------------------------------
                                        Total timers : 18
Closing OTF2 event files...
Calling reduce_node_properties on rank 0
reduce_node_properties hostname pop-os
reduce_node_properties hostname pop-os
Writing OTF2 definition files...
OTF2 Region : Continuation with executor
OTF2 Region : Continuation without executor
OTF2 Region : Dataflow with executor
OTF2 Region : Dataflow without executor
OTF2 Region : async_launch_policy_dispatch
OTF2 Region : hpx::actions::action<void (*)(hpx::components::server::runtime_support::call_shutdown_functions_action const&, std::vector<hpx::naming::id_type, std::allocator<hpx::naming::id_type> > const&, unsigned long, std::integral_constant<bool, true>, bool const&), &hpx::lcos::detail::broadcast_invoker<hpx::components::server::runtime_support::call_shutdown_functions_action, std::integral_constant<bool, true>, bool>::call, hpx::actions::detail::this_type>
OTF2 Region : hpx::agas::server::primary_namespace::colocate_action
OTF2 Region : hpx::components::server::runtime_support::call_shutdown_functions_action
OTF2 Region : hpx::components::server::runtime_support::load_components_action
OTF2 Region : hpx::components::server::runtime_support::shutdown_all_action
OTF2 Region : run_helper
Writing OTF2 Global definition file...
Writing OTF2 Node information...
Writing OTF2 Communicators...
Closing the archive...
```
You can see in the output - that after the region reduction, the tasks are all present and correct - but and vampir shows no dataflow or continuation tasks (amongst others)
![image](https://user-images.githubusercontent.com/941548/72670236-9a093e80-3a3b-11ea-94ba-2446ad1c311b.png)

As a sanity test, I rebuilt hpx with parcelport mpi on, networking on and my mpi fixes `#ifdef'd` out, I can confirm that a multi node normal HPX does not show the tasks either, so as I suspected, the apex support doesn't work on multinode jobs anyway.

Here is the output from 2 ranks running a small test - note the absence of the dataflow and continuation tasks in the vampir display.
```
Elapsed time: 0.908342 seconds
Cores detected: 8
Worker Threads observed: 4
Available CPU time: 3.63337 seconds

Timer                                                : #calls  |    mean  |   total  |  % total  
------------------------------------------------------------------------------------------------
                          Continuation with executor :        1      0.004      0.004      0.108
                       Continuation without executor :        1      0.001      0.001      0.029
                              Dataflow with executor :        1      0.001      0.001      0.035
                           Dataflow without executor :        1      0.006      0.006      0.162
                        async_launch_policy_dispatch :        3      0.000      0.000      0.004
                                     background_work :        5      0.000      0.000      0.011
            broadcast_call_shutdown_functions_action :        2      0.000      0.001      0.015
                      call_shutdown_functions_action :        3      0.001      0.004      0.109
                component_namespace_bind_name_action :       16      0.002      0.033      0.907
              component_namespace_bind_prefix_action :       17      0.001      0.015      0.401
                         dijkstra_termination_action :        4      0.001      0.004      0.119
                              load_components_action :        1      0.009      0.009      0.251
                      locality_namespace_free_action :        1      0.000      0.000      0.004
                   primary_namespace_colocate_action :        2      0.000      0.000      0.005
                      primary_namespace_route_action :        2      0.000      0.001      0.015
                                          run_helper :        1      0.026      0.026      0.702
                                 shutdown_all_action :        1      0.005      0.005      0.126
                        symbol_namespace_bind_action :        2      0.000      0.000      0.007
                    symbol_namespace_on_event_action :        1      0.000      0.000      0.008
                     symbol_namespace_resolve_action :        1      0.000      0.000      0.004
                            update_agas_cache_action :        1      0.000      0.000      0.002
                                           APEX Idle :                          3.524     96.978
------------------------------------------------------------------------------------------------
                                        Total timers : 67
```
here is the vampir screenshot 
![image](https://user-images.githubusercontent.com/941548/72669044-72ab7500-3a2d-11ea-8e89-1e69284744c1.png)

It might be that there is a problem with vampir? I'm using version 9.7 - I am attaching a tarball of the OTF archive just in case
[pause] = grrrr. can't add a tarball here. Will try to attach the file some other way....

**Comment by biddisco** [comment]
https://cloud.cscs.ch/owncloud/index.php/s/6QR9sDOa6Vx9rsW
should be the link to the otf tarball. If someone could try loading this in a different viewre than vampir 3.7 I'd be very grateful for info about whether it looks consisten with the apex screen output

**Comment by khuck** [comment]
@biddisco I am trying to put together an isolated APEX test to see whether it's an APEX issue when reducing the data.

**Comment by khuck** [comment]
@biddisco Can you commit this program to the same tests/unit/apex directory?  Which branch is this program in?

**Comment by biddisco** [comment]
One the dataflow-annotations branch in PR #4311 (merge that branch) there is an anotations test. If you run that on multiple ranks, it ought to be a reasonable test. Any simple test that creates a few annotations should show the same behaviour

**Comment by khuck** [comment]
That doesn't seem to work - there's no activity on any locality other than 0.  When I run that example with 2 mpi ranks and APEX_CSV_OUTPUT=1, I get this:

```console
[khuck@cyclops build]$ cat apex.0.csv 

"task","num calls","total cycles","total microseconds"
"1-Dataflow",1,0,19656
"1-Executor non-threaded Dataflow",1,0,1087
"1-Executor threaded Dataflow",1,0,1154
"1-apply Dataflow",1,0,18806
"1-async Dataflow",1,0,1179
"1-deferred Dataflow",1,0,1095
"1-fork Dataflow",1,0,1182
"2-Continuation",1,0,1082
"2-Executor non-threaded Continuation",1,0,1128
"2-Executor threaded Continuation",1,0,13181
"2-async Continuation",1,0,1102
"2-fork Continuation",1,0,1073
"<unknown>",1,0,13077
"async_launch_policy_dispatch",3,0,1770
"background_work",41,0,693
"broadcast_call_shutdown_functions_action",2,0,926
"call_shutdown_functions_action",3,0,9618
"hpx_main",3,0,3312
"load_components_action",1,0,20766
"primary_namespace_colocate_action",2,0,202
"run_helper",1,0,55936
"shutdown_all_action",1,0,68619

[khuck@cyclops build]$ cat apex.1.csv 

"task","num calls","total cycles","total microseconds"
"async_launch_policy_dispatch",2,0,2917
"background_work",40,0,672
"load_components_action",1,0,90665
"run_helper",1,0,36573

```

**Comment by biddisco** [comment]
It doesn't matter how many localities do work. What is important is whether the tasks appear in vampir or other tool traces. If other ranks doing nothing bothers you, then add `hpx:run-hpx-main` to the command line options.

**Comment by biddisco** [comment]
Is there anything I can do to move this along? I'm keen to get my patches on MPI+apex working on multinode jobs. Is there some additional test I can create that would help?

**Comment by khuck** [comment]
@biddisco sorry, I've been occupied with family stuff this week (traveling to attend a funeral this weekend), and an all-hands meeting next week.  I might not be able to look into it until after then.

**Comment by biddisco** [comment]
@khuck Please accept my condolences. In the meantime, it seems as though, if all ranks are doing work and the number of task types is limited, then traces are being generated correctly.
Here's 24 ranks doing 8 threads each - the output looks quite correct - at least this is good enough for me to use for this test (its a smaller test than the full cholesky, but fixing this will almost certainly tell me what to do with the others). (The only odd thing I can see initially, is that the overview track on each node doesn't display anything, but maybe it's not supposed to). 
![image](https://user-images.githubusercontent.com/941548/73518149-d6567a80-43fd-11ea-9137-d2c9f05a19e8.png)

**Comment by khuck** [comment]
@biddisco Thanks.  Can you share that trace so I can see what's going on with it?  I have vampir 9.7.1 installed.  The overview should show a summary of everything in the zoom window, so something strange is afoot.

**Comment by khuck** [comment]
tl/dr: I pushed some significant fixes and changes to APEX this past week regarding OTF2 support.  Please let me know if the current develop branch of APEX has any remaining unresolved problems. Thanks!

Specifically, the OTF2 thread, region and metric unification process had a bug where if any of those things had a space in the name (like a timer with a space in it), it wouldn't get resolved correctly for any rank other than 0 (and sometimes not even then).  OTF2 requires a global mapping at the end of execution, and that requires a reduction operation.  Unfortunately, distributed HPX implementations without MPI use the filesystem to do the unification.  It's been working, but it's not the best solution.  @weilewei is using this for tracing the HPX implementation of DCA++ and it's now working correctly for that case.

**Comment by biddisco** [comment]
Just bumping this to prevent the stale-bot from getting rid of it. I will test multi-node networking free apex and check that all is working, then close this if it is.

**Comment by khuck** [comment]
@biddisco any luck?
